File: macclean/cleaners/login_items.py
from pathlib import Path
import click
from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from macclean.core.utils import run_cmd

console = Console()


def _get_login_items_osascript() -> list[tuple[str, str]]:
    script = 'tell application "System Events" to get the properties of every login item'
    out, code = run_cmd(["osascript", "-e", script], timeout=10)
    if code != 0 or not out.strip():
        return []
    items = []
    for entry in out.split("},"):
        name, path = "", ""
        for part in entry.split(","):
            part = part.strip()
            if "name:" in part:
                name = part.split("name:")[-1].strip()
            elif "path:" in part:
                path = part.split("path:")[-1].strip()
        if name:
            items.append((name, path))
    return items


def _get_login_items_sfltool() -> list[tuple[str, str]]:
    out, code = run_cmd(["sfltool", "dumpbtm"], timeout=10)
    if code != 0:
        return []
    items = []
    name, url = "", ""
    for line in out.splitlines():
        line = line.strip()
        if "name =" in line:
            name = line.split("name =")[-1].strip().strip('"')
        elif "url =" in line:
            url = line.split("url =")[-1].strip().strip('"').replace("file://", "")
            if name:
                items.append((name, url))
                name, url = "", ""
    return items


@click.command("login-items")
@click.pass_context
def cmd(ctx):
    """Show apps that launch at login."""
    items = _get_login_items_sfltool()
    if not items:
        items = _get_login_items_osascript()

    if not items:
        console.print("[dim]No login items found (or permission denied).[/]")
        return

    table = Table(show_header=True, show_lines=True)
    table.add_column("App Name")
    table.add_column("Path")
    table.add_column("Exists?")

    for name, path in sorted(items, key=lambda x: x[0].lower()):
        exists = "[green]Yes[/]" if path and Path(path).exists() else "[red]No[/]" if path else "[dim]—[/]"
        table.add_row(name, path or "—", exists)

    console.print(Panel(table, title=f"[bold cyan]Login Items ({len(items)})[/]"))
    console.print("  [dim]To remove an item: System Settings → General → Login Items[/]")

File: src/cleaners/login_items.rs
use anyhow::Result;
use colored::Colorize;
use comfy_table::{Table, presets::UTF8_BORDERS_ONLY};
use std::path::PathBuf;
use crate::core::cmd::run_cmd;

/// Parse sfltool dumpbtm output.
/// Lines may look like:
///   name = "App Name"
///   url  = file:///Applications/App.app
fn parse_sfltool(output: &str) -> Vec<(String, String)> {
    let mut items: Vec<(String, String)> = Vec::new();
    let mut current_name = String::new();
    let mut current_url  = String::new();

    for line in output.lines() {
        let trimmed = line.trim();
        if let Some(rest) = trimmed.strip_prefix("name = ") {
            current_name = rest.trim_matches('"').to_string();
        } else if let Some(rest) = trimmed.strip_prefix("url = ") {
            current_url = rest
                .trim_matches('"')
                .trim_start_matches("file://")
                .to_string();
        }

        // When we have both, emit and reset
        if !current_name.is_empty() && !current_url.is_empty() {
            items.push((current_name.clone(), current_url.clone()));
            current_name.clear();
            current_url.clear();
        }
    }

    // Flush partial entry (name without url)
    if !current_name.is_empty() && current_url.is_empty() {
        items.push((current_name, String::new()));
    }

    items
}

pub fn run() -> Result<()> {
    println!("\n{}", "[ Login Items ]".cyan().bold());

    let mut items: Vec<(String, String)> = Vec::new();

    // Try sfltool first (macOS 13+)
    let sfltool_r = run_cmd(&["sfltool", "dumpbtm"]);
    if sfltool_r.success() && !sfltool_r.output.trim().is_empty() {
        items = parse_sfltool(&sfltool_r.output);
    }

    // Fallback: osascript
    if items.is_empty() {
        let osa_r = run_cmd(&[
            "osascript",
            "-e",
            "tell application \"System Events\" to get name of every login item",
        ]);
        if osa_r.success() && !osa_r.output.trim().is_empty() {
            for name in osa_r.output.split(',') {
                let n = name.trim().to_string();
                if !n.is_empty() {
                    items.push((n, String::new()));
                }
            }
        }
    }

    if items.is_empty() {
        println!("  No login items found (or not accessible).");
        return Ok(());
    }

    let mut table = Table::new();
    table.load_preset(UTF8_BORDERS_ONLY);
    table.set_header(vec!["App Name", "Path", "Exists?"]);

    for (name, path) in &items {
        let exists = if path.is_empty() {
            // Try to locate by name in common locations
            let candidates = [
                format!("/Applications/{}.app", name),
                format!("/Applications/{}", name),
                dirs::home_dir()
                    .map(|h| h.join("Applications").join(format!("{}.app", name))
                              .to_string_lossy().to_string())
                    .unwrap_or_default(),
            ];
            if candidates.iter().any(|c| !c.is_empty() && PathBuf::from(c).exists()) {
                "Yes".green().to_string()
            } else {
                "?".yellow().to_string()
            }
        } else {
            if PathBuf::from(path).exists() {
                "Yes".green().to_string()
            } else {
                "No".red().to_string()
            }
        };

        table.add_row(vec![
            name.clone(),
            if path.is_empty() { "N/A".to_string() } else { path.clone() },
            exists,
        ]);
    }
    println!("{}", table);

    Ok(())
}
